pmd-gherkin/src/main/antlr4/net/sourceforge/pmd/lang/gherkin/antlr4/Gherkin.g4

```diff
@@ -0,0 +1,118 @@
+grammar Gherkin;
+
+// PARSER
+
+main
+    // start comment needed because each comment should start on a new line except for the start comment
+    : STARTCOMMENT? feature description* instructionLine* NL* EOF
+    ;
+
+feature
+    : (NL* tagline)* NL* FEATURE?
+    ;
+
+instructionLine
+    : NL+ (instruction | datatable)
+    ;
+
+instruction
+    : rulex description* // the name "rule" is not allowed by ANTLR (used for internal usage), so calling it rulex
+    | stepInstruction description* (NL+ stepDescription description*)* (NL+ step)*
+    | tagline
+    | instructionDescription description*
+    ;
+
+stepInstruction
+    : background
+    | scenario
+    | scenarioOutline
+    ;
+
+background: BACKGROUND ;
+rulex: RULEX ;
+scenario: SCENARIO ;
+scenarioOutline : SCENARIOOUTLINE ;
+
+step : stepItem description*;
+
+stepItem
+    : and
+    | anystep
+    | but
+    | datatable
+    | given
+    | then
+    | when
+    | (NL* tagline )* NL* examples
+    ;
+
+tagline
+    : TAG+
+    ;
+
+and: AND ;
+anystep: ANYSTEP ;
+but: BUT ;
+datatable: DATATABLE+ ;
+given: GIVEN ;
+then: THEN ;
+when: WHEN ;
+examples: EXAMPLES ;
+
+// Descriptions
+instructionDescription: text | PARAMETER | AND | ANYSTEP | BUT | GIVEN | THEN | WHEN | SCENARIO ; // We have to deal with overlaps with keywords
+stepDescription: text | PARAMETER ; // We have to deal with overlaps with keywords
+description: text | PARAMETER | TAG | AND | ANYSTEP | BUT | DATATABLE | GIVEN | THEN | WHEN | SCENARIO | SCENARIOOUTLINE | STARTCOMMENT ; // We have to deal with overlaps with keywords
+
+text: TOKEN+ ;
+
+// LEXER
+
+// skipped
+
+BOMUTF8 : '\u00EF\u00BB\u00BF' -> skip ;
+
+BOMUTF16 : '\uFEFF' -> skip ;
+
+WHITESPACE: [ \t]+ -> channel(1) ;
+
+COMMENT: '\r'?'\n' [ \t]* '#' ~[\r\n]* -> channel(2) ;
+
+STARTCOMMENT: '#' ~[\r\n]* ;
+
+DOCSTRING1
+    : '"""' .*? '"""' ;
+
+DOCSTRING2
+    : '```' .*? '```' ;
+
+// Instructions
+BACKGROUND: 'Background:' ;
+EXAMPLES: ('Examples:' | 'Scenarios:') ;
+FEATURE: 'Feature:';
```

Comment:
We probably should mention somewhere, that only English is supported - non English variants of Gherkin can't be analyzed by pmd (see https://cucumber.io/docs/gherkin/reference/#spoken-languages)